### ISSUE TITLE
502 - Web server received an invalid response while acting as a gateway or proxy server.

### ISSUE THREAD

**Issue by shirhatti** [aspnet/KestrelHttpServer#1908]
_From @kimbell on June 8, 2017 6:51_

We have the current setup:
Web Browser (Swagger) -> Netflix Zuul gateway -> IIS -> Kestrel

In our development environment everything is working as expected; requests and responses make their way through the whole chain. When I deployed my application to another environment, things started going pear shaped. 

If I hit the server directly from en external browser, then things work fine. This confirms that the IIS -> Kestrel integration is working fine. 

When hitting through the gateway, I get the 502 error.
I enabled failed request logging, and found the following information. 
Something is going on between Zuul and IIS. I don't see anything in the .NET Core logging system.

```
MODULE_SET_RESPONSE_ERROR_STATUS 
ModuleName AspNetCoreModule 
Notification EXECUTE_REQUEST_HANDLER 
HttpStatus 502 
HttpReason Bad Gateway 
HttpSubStatus 3 
ErrorCode The parameter is incorrect.  (0x80070057) 

SET_RESPONSE_ERROR_DESCRIPTION 
ErrorDescription There was a connection error while trying to route the request.
```

Since I'm doing GET requests, I only have headers. 
```
Connection: Keep-Alive
Transfer-Encoding: chunked
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Encoding: gzip
Accept-Language: en-US,en;q=0.8
Host: stage2-cs02:5100
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36
upgrade-insecure-requests: 1
x-forwarded-host: stage2-gateway01.holding.intra:9000
x-forwarded-proto: http
x-forwarded-prefix: /checkin
x-forwarded-port: 9000
x-forwarded-for: 10.20.87.14
```

I fired up Fiddler and triggered a GET response directly against the server using these headers, and got the same 502 error. After eliminating each one, I was left with 

`Transfer-Encoding: chunked`

If I add that to a GET request, it fails. 
Is this behavior by design?

Since we have this setup working in another environment, I'll work with the guys that set up Zuul to remove this header. 

.NET Core 1.1.0 x64
Windows server 6.3.9600

_Copied from original issue: aspnet/AspNetCoreModule#106_

**Comment by shirhatti** [comment]
_From @kimbell on June 8, 2017 8:36_

The version of the IIS AspNetCore Module is 7.1.19720.0. 
Verified same version on both environments. 

Adding the header to request on my development environment (that works as expected) also triggers the same 502 error.

**Comment by shirhatti** [comment]
_From @kimbell on June 8, 2017 8:45_

Fired up Visual Studio and Kestrel om my local machine. If I trigger GET requests with the chunked transfer encoding, it gives me 200 status back, but reports an exception.

```
Connection id "0HL5E96E07SNN" bad request data: "Bad chunk size data."
Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Bad chunk size data.
```
I've never seen any traces of this in the logs when running Kestrel behind IIS.

**Comment by shirhatti** [comment]
_From @kimbell on June 8, 2017 10:5_

After pestering the Java guys that set up Zuul, they suddenly discovered they had a version mismatch between the environments. When they updated the problematic one, things started to work as expected. 

Figuring out what actually went wrong could have been a bit smoother. Once you configure failed request logging, mapping the error 'The parameter is incorrect' to a request header isn't straight forward.

**Comment by shirhatti** [comment]
Sounds like ANCM was transparently just forwarding the empty body chunked request to Kestrel. Off the top of my head, I'm not sure how Kestrel responds to such a request.

**Comment by halter73** [comment]
@shirhatti What should we do in Kestrel? Why did you move this issue from the ANCM to the Kestrel repo?

> When hitting through the gateway, I get the 502 error.
I enabled failed request logging, and found the following information.
Something is going on between Zuul and IIS. I don't see anything in the .NET Core logging system.

The above makes it sound like the request doesn't even make it to Kestrel.

**Comment by shirhatti** [comment]
I think further down in the thread, the OP clarified that the request made it to Kestrel, unless I misunderstood.

> Connection id "0HL5E96E07SNN" bad request data: "Bad chunk size data."
Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Bad chunk size data.

I interpreted this as how do we respond to an empty-body chunked request?

**Comment by kimbell** [comment]
When I sent the request directly against Kestrel, I got an error message, so Kestrel seems to be behaving correctly. The issue is when the request comes via IIS, then Kestrel never receives the request.

**Comment by halter73** [comment]
@kimbell A request with a `Transfer-Encoding: chunked` header and a completely empty body is invalid and should be rejected by any server that receives it.

It seems that you already recognize that Kestrel is doing the right thing by logging and rejecting request, but you still seem to think that IIS/ANCM should forwarding the same bad request. Am I understanding the issue correctly?

Not forwarding bad requests seems like the correct behavior to me. Maybe ANCM could log better and/or HTTP set a more appropriate status code.

**Comment by muratg** [comment]
What @halter73 said :)

**Comment by kimbell** [comment]
@halter73 @muratg 
This issue illustrates the amount of effort required to figure out what is going wrong when ANCM is stopping requests to Kestrel. This is just one example of a case; there are probably others out there as well. 

In order to figure things out, I needed to install Trace logging in IIS. Since this problem occurred in development, no big deal, but had it started in production it would have required additional steps to get things set up. 

The root cause of the problem was an external component sending bad requests, so it's not far fetched that an IIS and Kestrel setup working one day fails the next. And people will likely blame the IIS configuration since that is giving the error. 

When I eventually got trace logging set up, the error `The parameter is incorrect` doesn't give you a lot to work with. Linking this to an invalid header isn't directly obvious. 

If ANCM should behave 'smart' and not just forward every request to Kestrel, It would be nice to have better error reporting. Ideally the error message should indicate what header was incorrect. If that level of detail is difficult to achieve in ANCM, then I vote for just passing the error to Kestrel. Had that been the case, the problem would have been a lot easier to figure out.
